plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -256,534 +254,666 @@ struct block_time_tracker {
+         _producer_watermarks.emplace(producer, std::make_pair(block_num, timestamp));
+      }
+   }
+
+   std::optional<producer_watermark> get_watermark(account_name producer) const {
+      auto itr = _producer_watermarks.find(producer);
+
+      if (itr == _producer_watermarks.end())
+         return {};
+
+      return itr->second;
+   }
+
+   void on_block(const block_state_ptr& bsp) {
+      auto& chain  = chain_plug->chain();
+      auto  before = _unapplied_transactions.size();
+      _unapplied_transactions.clear_applied(bsp);
+      chain.get_mutable_subjective_billing().on_block(_log, bsp, fc::time_point::now());
+      if (before > 0) {
+         fc_dlog(_log,
+                 "Removed applied transactions before: ${before}, after: ${after}",
+                 ("before", before)("after", _unapplied_transactions.size()));
+      }
+   }
+
+   void on_block_header(const block_state_ptr& bsp) { consider_new_watermark(bsp->header.producer, bsp->block_num, bsp->block->timestamp); }
+
+   void on_irreversible_block(const signed_block_ptr& lib) {
+      const chain::controller& chain = chain_plug->chain();
+      EOS_ASSERT(chain.is_write_window(), producer_exception, "write window is expected for on_irreversible_block signal");
+      _irreversible_block_time = lib->timestamp.to_time_point();
+      _snapshot_scheduler.on_irreversible_block(lib, chain);
+   }
+
+   void abort_block() {
+      auto& chain = chain_plug->chain();
+
+      if (chain.is_building_block()) {
+         _time_tracker.report(_idle_trx_time, chain.pending_block_num());
       }
+      _unapplied_transactions.add_aborted(chain.abort_block());
+      _idle_trx_time = fc::time_point::now();
+   }
 
-      bool on_incoming_block(const signed_block_ptr& block, const std::optional<block_id_type>& block_id, const block_state_ptr& bsp) {
-         auto& chain = chain_plug->chain();
-         if ( _pending_block_mode == pending_block_mode::producing ) {
-            fc_wlog( _log, "dropped incoming block #${num} id: ${id}",
-                     ("num", block->block_num())("id", block_id ? (*block_id).str() : "UNKNOWN") );
-            return false;
-         }
+   bool on_incoming_block(const signed_block_ptr& block, const std::optional<block_id_type>& block_id, const block_state_ptr& bsp) {
+      auto& chain = chain_plug->chain();
+      if (_pending_block_mode == pending_block_mode::producing) {
+         fc_wlog(
+            _log, "dropped incoming block #${num} id: ${id}", ("num", block->block_num())("id", block_id ? (*block_id).str() : "UNKNOWN"));
+         return false;
+      }
 
-         // start a new speculative block, speculative start_block may have been interrupted
-         auto ensure = fc::make_scoped_exit([this](){
-            schedule_production_loop();
-         });
+      // start a new speculative block, speculative start_block may have been interrupted
+      auto ensure = fc::make_scoped_exit([this]() { schedule_production_loop(); });
```

Comment:
Interesting that it forces `if` to two lines but one line functions to one line.